src/workers/parser/mapBindings.js

```diff
@@ -10,20 +10,51 @@ import { isTopLevel } from "./utils/helpers";
 import generate from "@babel/generator";
 import * as t from "@babel/types";
 
+function getIdentifierName(identifier, bindings) {
+  return bindings.includes(identifier.name)
+    ? identifier
+    : t.memberExpression(t.identifier("self"), identifier);
+}
+
+export function globalizeObjectDestructuring(node, bindings) {
+  for (const property of node.properties) {
+    if (t.isIdentifier(property.value)) {
+      property.value = getIdentifierName(property.value, bindings);
+    } else if (
+      t.isAssignmentPattern(property.value) &&
+      t.isIdentifier(property.value.left)
+    ) {
+      property.value.left = getIdentifierName(property.value.left, bindings);
+    }
+  }
+}
+
 // translates new bindings `var a = 3` into `self.a = 3`
 // and existing bindings `var a = 3` into `a = 3` for re-assignments
 function globalizeDeclaration(node, bindings) {
   return node.declarations.map(declaration => {
-    const identifier = bindings.includes(declaration.id.name)
-      ? declaration.id
-      : t.memberExpression(t.identifier("self"), declaration.id);
+    if (t.isPattern(declaration.id)) {
+      return t.expressionStatement(
+        t.assignmentExpression("=", declaration.id, declaration.init)
+      );
+    }
+
+    const identifier = getIdentifierName(declaration.id, bindings);
 
     return t.expressionStatement(
       t.assignmentExpression("=", identifier, declaration.init)
     );
   });
 }
 
+function globalizeArrayDestructuring(node, bindings) {
+  for (const [i, element] of node.elements.entries()) {
+    if (t.isIdentifier(element)) {
+      node.elements[i] = getIdentifierName(element, bindings);
+    }
```

Comment:
Looks like this also needs to handle the `AssignmentPattern` case.

Reply:
I think it's handled now. Correct me if I'm wrong.